diff -up fakenes-0.5.9-beta3/default.cbd~ fakenes-0.5.9-beta3/default.cbd
--- fakenes-0.5.9-beta3/default.cbd~	2007-06-25 18:43:10.000000000 +0200
+++ fakenes-0.5.9-beta3/default.cbd	2011-12-10 20:37:13.528058876 +0100
@@ -297,7 +297,7 @@ ifopt clean goto clean
 CXXFLAGS = "${CFLAGS}"
 
 # We have to include this or we get undefined references to new & delete.
-LDFLAGS += ' -lstdc++'
+LDFLAGS += ' -lstdc++ -lm'
 
 compile ${SRCFILES}
 
