--- fakenes-0.5.9-beta3/src/audio.cpp~	2007-06-22 04:26:25.000000000 +0200
+++ fakenes-0.5.9-beta3/src/audio.cpp	2009-10-25 10:45:45.000000000 +0100
@@ -7,6 +7,7 @@
    You must read and accept the license prior to use. */
 
 #include <allegro.h>
+#include <cstdio>
 #include <cstdlib>
 #include <cstring>
 #include <vector>
--- fakenes-0.5.9-beta3/src/audio.cpp~	2009-10-25 10:50:09.000000000 +0100
+++ fakenes-0.5.9-beta3/src/audio.cpp	2009-10-25 10:50:40.000000000 +0100
@@ -6,14 +6,16 @@
    This is free software.  See 'LICENSE' for details.
    You must read and accept the license prior to use. */
 
+/* This includes c++ <vector> which uses min and max internally, which get
+   defined as macros in allegro.h, so include this before allegro.h */
+#include "audio_int.h"
 #include <allegro.h>
 #include <cstdio>
 #include <cstdlib>
 #include <cstring>
 #include <vector>
 #include "apu.h"
 #include "audio.h"
-#include "audio_int.h"
 #include "audiolib.h"
 #include "common.h"
 #include "debug.h"
--- fakenes-0.5.9-beta3/src/audiolib.cpp~	2007-06-16 21:27:27.000000000 +0200
+++ fakenes-0.5.9-beta3/src/audiolib.cpp	2009-10-25 10:51:01.000000000 +0100
@@ -6,11 +6,13 @@
    This is free software.  See 'LICENSE' for details.
    You must read and accept the license prior to use. */
 
+/* This includes c++ <vector> which uses min and max internally, which get
+   defined as macros in allegro.h, so include this before allegro.h */
+#include "audio_int.h"
 #include <allegro.h>
 #include <cstdlib>
 #include <cstring>
 #include "audio.h"
-#include "audio_int.h"
 #include "audiolib.h"
 #include "common.h"
 #include "debug.h"
