--- fakenes-0.5.8/src/audiolib.c.alutinit	2006-04-24 23:21:22.000000000 +0200
+++ fakenes-0.5.8/src/audiolib.c	2006-08-13 00:41:18.000000000 +0200
@@ -415,8 +415,13 @@
    /* Interpolation is not supported. */
    audio_interpolation = FALSE;
 
-   alutInit (&saved_argc, saved_argv);
-   AL_CHECK();
+   if(!alutInit (&saved_argc, saved_argv))
+   {
+      /* make sure other openal functions don't get called as that causes
+         crashes */
+      audiolib_driver = NULL; 
+      return (1);
+   }
 
    /* Autodetect settings. */
 
