--- a/src/nsf.cpp
+++ b/src/nsf.cpp
@@ -410,11 +410,11 @@
          clear_to_color(video_buffer, ppu_get_background_color());
 
          textprintf_ex(video_buffer, small_font, 8,     8 + (12 * 0), 1 + 0x30, -1, "TITLE:");
-         textprintf_ex(video_buffer, small_font, 8 + 8, 8 + (12 * 1), 1 + 0x30, -1, (const char *)&nsf.name[0]);
+         textprintf_ex(video_buffer, small_font, 8 + 8, 8 + (12 * 1), 1 + 0x30, -1, "%s", (const char *)&nsf.name[0]);
          textprintf_ex(video_buffer, small_font, 8,     8 + (12 * 2), 1 + 0x30, -1, "COMPILED BY:");
-         textprintf_ex(video_buffer, small_font, 8 + 8, 8 + (12 * 3), 1 + 0x30, -1, (const char *)&nsf.artist[0]);
+         textprintf_ex(video_buffer, small_font, 8 + 8, 8 + (12 * 3), 1 + 0x30, -1, "%s", (const char *)&nsf.artist[0]);
          textprintf_ex(video_buffer, small_font, 8,     8 + (12 * 4), 1 + 0x30, -1, "COPYRIGHT:");
-         textprintf_ex(video_buffer, small_font, 8 + 8, 8 + (12 * 5), 1 + 0x30, -1, (const char *)&nsf.copyright[0]);
+         textprintf_ex(video_buffer, small_font, 8 + 8, 8 + (12 * 5), 1 + 0x30, -1, "%s", (const char *)&nsf.copyright[0]);
 
          textprintf_ex(video_buffer, small_font, 8, 8 + (12 * 7), 1 + 0x30, -1, "Track %d of %d", currentSong, nsf.totalSongs);
          textprintf_ex(video_buffer, small_font, 8, 8 + (12 * 8) + 2, 1 + 0x30, -1, "Press A for next track");

